### ISSUE TITLE
[FEATURE] More Robust Schema Discrimination in Models with multiple attached Schemas

### ISSUE THREAD

**Issue by abraxas** [dynamoose/dynamoose#1014]
### Summary:

When Models support multiple schemas, the schema match is applied prior to validation, which means you can get a ValidationException when a model would have been valid.  This is a feature request, not a bug, because the docs do not currently say this coding pattern is viable.

#### Schema
```
const BaseSchema = dynamoose.schema({
  dataId: {
    type: String,
    hashKey: true,
  },
  type: String //Alternatively, an enum type here...
});

const OrderSchema = dynamoose.schema({
  ...BaseSchema,
  type: {
    validate: "order"
  }
});

const CustomerSchema = dynamoose.schema({
  ...BaseSchema,
  type: {
    validate: "customer"
  }
});

```

#### Model
```
const Model = DataModel = dynamoose.model('table_name',[OrderSchema,CustomerSchema]);
```

#### General
```
const results = await DataModel.query({'dataId': {eq: 'SOME_ID'} }).all().exec();
// Unless the model match is unambiguous, this above line sometimes results in a validation exception 
//    depending on the value of "type" as compared to the Schema it decides to match
```

### Environment:

Operating System: any
Operating System Version: any
Node.js version (`node -v`): 12.13.1
NPM version: (`npm -v`): 6.12.1
Dynamoose version: 2.3.0-beta.1

### Other information (if applicable):

We had a discussion in Slack about this, and there are a few takeaways.

1. We probably want to consider `enum` and `required` as well as `validates`.  
2. We need to consider speed.  Validation can be slow, so validating all schemas before matching one might be sub-optimal.
3. I suggested that we can RE-validate in the unique case of a Model with multiple Schemas when a validation fails.  This would be somewhat slow, but need only apply in that one situation, where it would generally be desirable
4. To short-circuit re-validation we can have a boolean flag called "discriminator" on fields that will always apply when trying to pick a Schema, or just support a specific matchable `_type` parameter in the schema (with an optional configuration for alternate typefield name).  
5. Alternatively, also support a `usingSchema` which mimics the behavior of `using` but for the Schema instead of the index.

For the #4 idea, consider:

```
const BaseSchema = dynamoose.schema({
  dataId: {
    type: String,
    hashKey: true,
  },
  type: String
});

const OrderSchema = dynamoose.schema({
  ...BaseSchema,
}, {
  _type: 'order',
  typeField: 'type',
});

const CustomerSchema = dynamoose.schema({
  ...BaseSchema,
}, {
  _type: 'customer',
  typeField: 'type',
});
```

### Other:
- [X] I have read through the Dynamoose documentation before posting this issue
- [X] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this feature has not already been suggested before
- [X] I have filled out all fields above
- [X] I am running the latest version of Dynamoose

**Comment by abraxas** [comment]
From @fishcharlie from Slack:

> Just for reference, https://github.com/dynamoose/dynamoose/blob/94b6295a56151d8ae5debf350cbcdc248342a4da/lib/Document.ts#L285. That is where for each document it’s determining the correct schema to use.
> If you follow the path down for those functions, you end up at this method. https://github.com/dynamoose/dynamoose/blob/94b6295a56151d8ae5debf350cbcdc248342a4da/lib/Schema.ts#L376. With a little bit of information about the ranking system (https://github.com/dynamoose/dynamoose/blob/94b6295a56151d8ae5debf350cbcdc248342a4da/lib/Schema.ts#L412-L414).
> Now, validation happens at https://github.com/dynamoose/dynamoose/blob/94b6295a56151d8ae5debf350cbcdc248342a4da/lib/Document.ts#L414-L435.
> So ideally, we wouldn’t do it twice. But figuring out how to refactor that to be efficient, might be the biggest challenge.
> One more note. We do need to think about enum and possibly required, and some of the other property settings.

**Comment by abraxas** [comment]
To keep things simple+fast, I'm going to start by implementing bullet point 4.

I will be using the fields `modelType` and `modelTypeFieldName` to keep them unambiguous.  They will add a weight of 0.25 on failure, so a type mismatch still beats an absolute failure, but loses to a non-mismatch with incorrect types.

**Comment by fishcharlie** [comment]
Closing this issue since it was fixed in #1089.
